gematria/datasets/basic_block_utils.cc

```diff
@@ -0,0 +1,93 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gematria/datasets/basic_block_utils.h"
+
+#include <map>
+
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "X86RegisterInfo.h"
+#include "llvm/MC/MCInstrInfo.h"
+#include "llvm/MC/MCRegisterInfo.h"
+
+using namespace llvm;
+
+namespace gematria {
+
+unsigned getSuperRegister(unsigned OriginalRegister,
+                          const MCRegisterInfo &RegisterInfo) {
+  unsigned SuperRegister = OriginalRegister;
+  for (MCPhysReg CurrentSuperRegister :
+       RegisterInfo.superregs_inclusive(OriginalRegister)) {
+    SuperRegister = CurrentSuperRegister;
+  }
+  // Only return super registers for GPRs. Since mainly GPRs will be used for
+  // addressing, redefining other aliasing registers (like vector registers)
+  // does not matter as much.
+  if (RegisterInfo.getRegClass(X86::GR64RegClassID).contains(SuperRegister))
+    return SuperRegister;
+  else
+    return OriginalRegister;
+}
+
+std::vector<unsigned> BasicBlockUtils::getUsedRegisters(
+    const std::vector<DisassembledInstruction> &Instructions,
+    const MCRegisterInfo &RegisterInfo, const MCInstrInfo &InstructionInfo) {
+  std::map<unsigned, bool> UsedInputRegisters;
+  std::map<unsigned, bool> DefinedInBBRegisters;
+  for (const gematria::DisassembledInstruction &Instruction : Instructions) {
+    unsigned InstructionDefs =
+        InstructionInfo.get(Instruction.mc_inst.getOpcode()).getNumDefs();
+    for (unsigned OperandIndex = InstructionDefs;
+         OperandIndex < Instruction.mc_inst.getNumOperands(); ++OperandIndex) {
+      if (Instruction.mc_inst.getOperand(OperandIndex).isReg()) {
+        unsigned RegisterNumber =
+            Instruction.mc_inst.getOperand(OperandIndex).getReg();
+        if (RegisterNumber == 0) continue;
+        unsigned SuperRegister = getSuperRegister(RegisterNumber, RegisterInfo);
+        // If the register was already defined within the basic block, we don't
+        // need to define it ourselves.
+        if (DefinedInBBRegisters.count(SuperRegister) > 0) continue;
+        UsedInputRegisters[SuperRegister] = true;
+      }
+    }
+    // We also need to handle instructions that have implict uses.
+    for (unsigned ImplicitlyUsedRegister :
+         InstructionInfo.get(Instruction.mc_inst.getOpcode()).implicit_uses()) {
+      UsedInputRegisters[getSuperRegister(ImplicitlyUsedRegister,
+                                          RegisterInfo)] = true;
+    }
```

Comment:
We also need to track implicit defs of the register.

Reply:
Done. Test added with `rdtsc`.